File: cpp/tests/test_live_data_map_parser.cpp
// SPDX-License-Identifier: MIT
//
// doctest unit tests for tuner_core::live_data_map_parser.

#include "doctest.h"

#include "tuner_core/live_data_map_parser.hpp"

#include <string>

using namespace tuner_core::live_data_map_parser;

// ---------------------------------------------------------------------------
// ChannelEncoding helpers
// ---------------------------------------------------------------------------

TEST_CASE("parse_encoding handles all known encodings") {
    CHECK(parse_encoding("U08") == ChannelEncoding::U08);
    CHECK(parse_encoding("U08 bits") == ChannelEncoding::U08_BITS);
    CHECK(parse_encoding("U16 LE") == ChannelEncoding::U16_LE);
    CHECK(parse_encoding("S16 LE") == ChannelEncoding::S16_LE);
    CHECK(parse_encoding("U32 LE") == ChannelEncoding::U32_LE);
}

TEST_CASE("parse_encoding is case-insensitive") {
    CHECK(parse_encoding("u08") == ChannelEncoding::U08);
    CHECK(parse_encoding("u16 le") == ChannelEncoding::U16_LE);
}

TEST_CASE("parse_encoding falls through to UNKNOWN") {
    CHECK(parse_encoding("garbage") == ChannelEncoding::UNKNOWN);
    CHECK(parse_encoding("") == ChannelEncoding::UNKNOWN);
}

TEST_CASE("byte_width matches the documented widths") {
    CHECK(byte_width(ChannelEncoding::U08) == 1);
    CHECK(byte_width(ChannelEncoding::U08_BITS) == 1);
    CHECK(byte_width(ChannelEncoding::U16_LE) == 2);
    CHECK(byte_width(ChannelEncoding::S16_LE) == 2);
    CHECK(byte_width(ChannelEncoding::U32_LE) == 4);
    CHECK(byte_width(ChannelEncoding::UNKNOWN) == 0);
}

// ---------------------------------------------------------------------------
// parse_text — full header parse
// ---------------------------------------------------------------------------

const char* SAMPLE_HEADER = R"(
/*
 * byte    ridx  field                          encoding       notes
 * 4-5      4    MAP                            U16 LE         map (kPa)
 * 14-15    13   RPM                            U16 LE         rpm
 * 84       57   status3                        U08 bits       status3 [LOCKED]
 * 100      -    AEamount >> 1                  U08            AEamount [low]
 */

#define LIVE_DATA_MAP_SIZE  148U

static constexpr uint16_t OCH_OFFSET_RUNTIME_STATUS_A    = 147U;
static constexpr uint16_t OCH_OFFSET_BOARD_CAPABILITY_FLAGS = 130U;
static constexpr uint16_t OCH_OFFSET_FLASH_HEALTH_STATUS = 131U;
)";

TEST_CASE("parse_text extracts log_entry_size") {
    auto c = parse_text(SAMPLE_HEADER);
    CHECK(c.log_entry_size == 148);
}

TEST_CASE("parse_text extracts the OCH_OFFSET constants") {
    auto c = parse_text(SAMPLE_HEADER);
    REQUIRE(c.runtime_status_a_offset.has_value());
    CHECK(*c.runtime_status_a_offset == 147);
    REQUIRE(c.board_capability_flags_offset.has_value());
    CHECK(*c.board_capability_flags_offset == 130);
    REQUIRE(c.flash_health_status_offset.has_value());
    CHECK(*c.flash_health_status_offset == 131);
}

TEST_CASE("parse_text extracts entries from the doxygen table") {
    auto c = parse_text(SAMPLE_HEADER);
    REQUIRE(c.entries.size() == 4);

    // MAP
    CHECK(c.entries[0].name == "map");  // notes-derived name
    CHECK(c.entries[0].byte_start == 4);
    CHECK(c.entries[0].byte_end == 5);
    CHECK(c.entries[0].readable_index.value() == 4);
    CHECK(c.entries[0].encoding == ChannelEncoding::U16_LE);
    CHECK(c.entries[0].field == "MAP");
    CHECK(c.entries[0].width() == 2);
    CHECK_FALSE(c.entries[0].locked);

    // RPM
    CHECK(c.entries[1].name == "rpm");
    CHECK(c.entries[1].byte_start == 14);
    CHECK(c.entries[1].byte_end == 15);

    // status3 [LOCKED]
    CHECK(c.entries[2].name == "status3");
    CHECK(c.entries[2].locked);
    CHECK(c.entries[2].notes == "status3");  // [LOCKED] stripped
    CHECK(c.entries[2].encoding == ChannelEncoding::U08_BITS);

    // single-byte row with '-' readable_index
    CHECK(c.entries[3].byte_start == 100);
    CHECK(c.entries[3].byte_end == 100);
    CHECK_FALSE(c.entries[3].readable_index.has_value());
    CHECK(c.entries[3].encoding == ChannelEncoding::U08);
}

TEST_CASE("parse_text accepts a firmware_signature pass-through") {
    auto c = parse_text(SAMPLE_HEADER, std::string("speeduino 202501-T41"));
    REQUIRE(c.firmware_signature.has_value());
    CHECK(*c.firmware_signature == "speeduino 202501-T41");
}

TEST_CASE("parse_text empty input yields empty contract") {
    auto c = parse_text("");
    CHECK(c.entries.empty());
    CHECK(c.log_entry_size == 0);
    CHECK_FALSE(c.runtime_status_a_offset.has_value());
}

TEST_CASE("parse_text skips DEPRECATED notes and uses field as name") {
    const char* header =
        " * 200      -    legacyVar                      U08            DEPRECATED: use newVar\n";
    auto c = parse_text(header);
    REQUIRE(c.entries.size() == 1);
    // Notes start with "DEPRECATED:" → name falls back to the field text.
    CHECK(c.entries[0].name == "legacyVar");
}

File: tests/unit/test_live_data_map_parser.py
"""Tests for the live_data_map.h parser and the resulting ChannelContract.

Future Phase 12 deliverable 4. Exercises the parser against:
  - synthetic header rows that lock the row format
  - the **real** firmware header at
    ``C:/Users/Cornelio/Desktop/speeduino-202501.6/speeduino/live_data_map.h``
    so any future schema bump that breaks our parser fails loudly
"""
from __future__ import annotations

import textwrap
from pathlib import Path

import pytest

from tuner.domain.channel_contract import (
    ChannelContract,
    ChannelEncoding,
    ChannelEntry,
)
from tuner.services.live_data_map_parser import LiveDataMapParser


_REAL_HEADER = Path(
    "C:/Users/Cornelio/Desktop/speeduino-202501.6/speeduino/live_data_map.h"
)


# ---------------------------------------------------------------------------
# Synthetic-row parser tests (no firmware tree dependency)
# ---------------------------------------------------------------------------

_SYNTHETIC = textwrap.dedent("""\
    /**
     * Byte(s)  RIdx  Field                          Encoding       Notes
     * -------- ----- ------------------------------ -------------- ---------------------------
     * 0        0     secl                           U08            secl
     * 4-5      4     MAP                            U16 LE         map (kPa)
     * 14-15    13    RPM                            U16 LE         RPM
     * 22-23    19    tpsDOT                         S16 LE         tpsDOT
     * 84       57    status3                        U08 bits       status3 [LOCKED]
     * 143-146  102   startRevolutions [31:0] LE     U32 LE         startRevolutions (Phase 10) [LOCKED]
     * 147      104   runtimeStatusA packed          U08 bits       runtimeStatusA (Phase 10) [LOCKED]
     */

    #define LIVE_DATA_MAP_SIZE  148U

    static constexpr uint16_t OCH_OFFSET_BOARD_CAPABILITY_FLAGS = 130U;
    static constexpr uint16_t OCH_OFFSET_FLASH_HEALTH_STATUS    = 131U;
    static constexpr uint16_t OCH_OFFSET_RUNTIME_STATUS_A       = 147U;
""")


class TestSyntheticParser:
    def _parse(self) -> ChannelContract:
        return LiveDataMapParser().parse_text(_SYNTHETIC)

    def test_log_entry_size_extracted(self) -> None:
        contract = self._parse()
        assert contract.log_entry_size == 148

    def test_special_offsets_extracted(self) -> None:
        contract = self._parse()
        assert contract.runtime_status_a_offset == 147
        assert contract.board_capability_flags_offset == 130
        assert contract.flash_health_status_offset == 131

    def test_single_byte_row(self) -> None:
        contract = self._parse()
        secl = contract.find("secl")
        assert secl is not None
        assert secl.byte_start == 0
        assert secl.byte_end == 0
        assert secl.encoding == ChannelEncoding.U08
        assert secl.locked is False

    def test_two_byte_le_row(self) -> None:
        contract = self._parse()
        rpm = contract.find("RPM")
        assert rpm is not None
        assert rpm.byte_start == 14
        assert rpm.byte_end == 15
        assert rpm.encoding == ChannelEncoding.U16_LE
        assert rpm.width == 2

    def test_four_byte_u32_row(self) -> None:
        contract = self._parse()
        sr = contract.find("startRevolutions")
        assert sr is not None
        assert sr.byte_start == 143
        assert sr.byte_end == 146
        assert sr.encoding == ChannelEncoding.U32_LE
        assert sr.width == 4
        assert sr.locked is True

    def test_signed_row(self) -> None:
        contract = self._parse()
        tps = contract.find("tpsDOT")
        assert tps is not None
        assert tps.encoding == ChannelEncoding.S16_LE

    def test_bits_row_and_locked_flag(self) -> None:
        contract = self._parse()
        s3 = contract.find("status3")
        assert s3 is not None
        assert s3.encoding == ChannelEncoding.U08_BITS
        assert s3.locked is True
        # The [LOCKED] tag is stripped from the notes column.
        assert "[LOCKED]" not in s3.notes

    def test_find_by_byte(self) -> None:
        contract = self._parse()
        # Bytes 4 and 5 both belong to MAP
        for byte in (4, 5):
            entry = contract.find_by_byte(byte)
            assert entry is not None
            assert entry.name == "map"
        assert contract.find_by_byte(99999) is None

    def test_unknown_row_skipped(self) -> None:
        text = textwrap.dedent("""\
            /**
             * Byte(s)  RIdx  Field                          Encoding       Notes
             * -------- ----- ------------------------------ -------------- ---------------------------
             * not a real row at all
             *          103   startRevolutions [31:16]       -              (readable hi-half of above)
             */
            #define LIVE_DATA_MAP_SIZE 148U
            """)
        contract = LiveDataMapParser().parse_text(text)
        assert contract.entries == ()
        assert contract.log_entry_size == 148


# ---------------------------------------------------------------------------
# Real firmware header — locks the parser to the actual file on disk
# ---------------------------------------------------------------------------

@pytest.mark.skipif(
    not _REAL_HEADER.exists(),
    reason="Speeduino firmware tree not on this machine",
)
class TestRealHeader:
    def _parse(self) -> ChannelContract:
        return LiveDataMapParser().parse(
            _REAL_HEADER, firmware_signature="speeduino 202501-T41",
        )

    def test_log_entry_size_is_148(self) -> None:
        contract = self._parse()
        assert contract.log_entry_size == 148

    def test_firmware_signature_attached(self) -> None:
        contract = self._parse()
        assert contract.firmware_signature == "speeduino 202501-T41"

    def test_runtime_status_a_at_offset_147(self) -> None:
        contract = self._parse()
        assert contract.runtime_status_a_offset == 147
        assert contract.board_capability_flags_offset == 130
        assert contract.flash_health_status_offset == 131

    def test_known_channels_present(self) -> None:
        """Spot-check that the parser caught the high-leverage rows.
        These names match the firmware notes column verbatim."""
        contract = self._parse()
        for name in ("secl", "map", "RPM", "tpsDOT", "status3", "knockCount"):
            assert contract.find(name) is not None, f"missing: {name}"

    def test_runtime_status_a_byte_is_a_locked_bits_field(self) -> None:
        contract = self._parse()
        entry = contract.find_by_byte(147)
        assert entry is not None
        assert entry.encoding == ChannelEncoding.U08_BITS
        assert entry.locked is True

    def test_start_revolutions_spans_4_bytes_and_is_locked(self) -> None:
        contract = self._parse()
        entry = contract.find_by_byte(143)
        assert entry is not None
        assert entry.byte_start == 143
        assert entry.byte_end == 146
        assert entry.encoding == ChannelEncoding.U32_LE
        assert entry.locked is True

    def test_total_byte_coverage_matches_log_entry_size(self) -> None:
        """Sum of all parsed entry widths should equal LIVE_DATA_MAP_SIZE.
        This is the structural canary that catches a parser regression
        the moment a real firmware row stops matching."""
        contract = self._parse()
        covered = sum(e.width for e in contract.entries)
        assert covered == contract.log_entry_size, (
            f"covered {covered} bytes, expected {contract.log_entry_size}"
        )
